apps/teams-test-app/src/components/DialogAPIs.tsx

```diff
@@ -1,90 +1,125 @@
+      name: 'sendMessageToChild',
+      title: 'sendMessageToChild',
+      onClick: async message => {
+        if (childWindowRef.current && childWindowRef.current !== null) {
+          const childWindow = childWindowRef.current;
+          await childWindow.postMessage(message);
+          return Promise.resolve('Message sent to child');
+        } else {
+          return Promise.resolve('childWindow doesnt exist');
+        }
+      },
+    });
 
-  const submitDialogWithInput = (submitDialogInput: string): void => {
-    if (submitDialogInput.length == 0) {
-      dialog.submit();
-      setSubmitRes('Teams client SDK call dialog.submit was called with no arguments');
-    } else {
-      try {
-        const parsedInput = JSON.parse(submitDialogInput);
-        dialog.submit(parsedInput.result, parsedInput.appIds);
-        setSubmitRes('Teams client SDK call dialog.submit was called with arguments');
-      } catch (error) {
-        if (error instanceof SyntaxError) {
-          setSubmitRes(generateJsonParseErrorMsg());
-        } else if (error instanceof Error) {
-          setSubmitRes(error.message);
+  const SendMessageToParent = (): ReactElement =>
+    ApiWithTextInput<string>({
+      name: 'sendMessageToParent',
+      title: 'sendMessageToParent',
+      onClick: async message => {
+        const parentWindow = ParentAppWindow.Instance;
+        if (parentWindow) {
+          await parentWindow.postMessage(message);
+          return Promise.resolve('Message sent to parent');
         } else {
-          setSubmitRes(JSON.stringify(error));
+          return Promise.resolve('parentWindow doesnt exist');
         }
-      }
-    }
-  };
+      },
+    });
 
-  const checkDialogCapability = (): void => {
-    if (dialog.isSupported()) {
-      setCapabilityCheckRes('Dialog module is supported');
-    } else {
-      setCapabilityCheckRes('Dialog module is not supported');
-    }
-  };
+  const RegisterForParentMessage = (): ReactElement =>
+    ApiWithoutInput({
+      name: 'registerForParentMessage',
+      title: 'registerForParentMessage',
+      onClick: async setResult => {
+        const parentWindow = ParentAppWindow.Instance;
+        parentWindow.addEventListener('message', (message: string) => {
+          setResult(message);
+        });
+        return Promise.resolve('Completed');
+      },
+    });
+
+  const SubmitDialogWithInput = (): ReactElement =>
+    ApiWithTextInput<{ result?: string | object; appIds?: string | string[] }>({
+      name: 'dialogSubmitWithInput',
+      title: 'Dialog Submit With Input',
+      onClick: {
+        validateInput: input => {},
+        submit: async submitInput => {
+          dialog.submit(submitInput.result, submitInput.appIds);
+          return '';
+        },
+      },
+    });
+
+  const CheckDialogCapability = (): ReactElement =>
+    ApiWithoutInput({
+      name: 'checkCapabilityDialog',
+      title: 'Check Capability Dialog',
+      onClick: async => {
```

Comment:
is this `async` making this delegate asynchronous (in which case you don't need to wrap your results in `Promise.resolve` (and I would expect TypeScript of TSLint to complain, because you're not using `await` anywhere in the body of the function) or is it treated as a parameter name (there's the `setResult` thing you could take here, isn't there?)